Add RedisModule_GetKeyNameFromIO()

yossigo opened the pull request:
A first go at exposing key names to rdb_save and rdb_load data type callbacks.  Seems like these should also be available for RESTORE and DUMP command handling, although my main focus was RDB processing and I may have overlooked some implications of those scenarios.

yossigo commented:
Have some postscript thoughts about whether the returned key should be a const or not.
Making it non-const would make it possible to do RetainString() and use it efficiently, but then the key should not be provided in the `DUMP` and `RESTORE` cases.

antirez commented:
@yossigo this no longer applies but reading the diff looks very reasonable to me. Only thing I wonder if the key obtained in this way is safe, for the module, to *retain*. Also this looks completely safe, AFAIK we'll never store such key object inside the data space, nor we create such objects, if I remember correctly, within the stack as an optimization or other magics. So I guess this is ready to merge after rebasing.

yossigo commented:
Hi @antirez, rebased this and also moved it out of experimental features.  About the const, I think there's really little if any value in retaining the key by the module.

antirez commented:
Thank you @yossigo, merged. Should I back port this into Redis 5?
